homeassistant/components/dormakaba_dkey/__init__.py

```diff
@@ -0,0 +1,125 @@
+
+    lock = DKEYLock(ble_device)
+    lock.set_association_data(
+        AssociationData.from_json(entry.data[CONF_ASSOCIATION_DATA])
+    )
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        lock.set_ble_device_and_advertisement_data(
+            service_info.device, service_info.advertisement
+        )
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
+    )
+
+    async def _async_update() -> None:
+        """Update the device state."""
+        try:
+            await lock.update()
+            await lock.disconnect()
+        except DKEY_EXCEPTIONS as ex:
+            raise UpdateFailed(str(ex)) from ex
+
+    startup_event = asyncio.Event()
+    cancel_first_update = lock.register_callback(lambda *_: startup_event.set())
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=lock.name,
+        update_method=_async_update,
+        update_interval=timedelta(seconds=UPDATE_SECONDS),
+    )
+
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except ConfigEntryNotReady:
+        cancel_first_update()
+        raise
+
+    try:
+        async with async_timeout.timeout(DEVICE_TIMEOUT):
+            await startup_event.wait()
+    except asyncio.TimeoutError as ex:
+        raise ConfigEntryNotReady(
+            "Unable to communicate with the device; "
+            f"Try moving the Bluetooth adapter closer to {lock.name}"
+        ) from ex
+    finally:
+        cancel_first_update()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = DormakabaDkeyData(
+        entry.title, lock, coordinator
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+
+    async def _async_stop(event: Event) -> None:
+        """Close the connection."""
+        await lock.disconnect()
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop)
+    )
+    return True
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    data: DormakabaDkeyData = hass.data[DOMAIN][entry.entry_id]
+    if entry.title != data.title:
+        await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        data: DormakabaDkeyData = hass.data[DOMAIN].pop(entry.entry_id)
```

Comment:
Observation: The naming throws me off here, data I would expect be some data description, not a class with connection properties.

Reply:
The class is a container of data which the integration stores in hass.data[], what would be a better name?
DkeyConfigEntryData?